lib/cluster/election.c

```diff
@@ -388,6 +388,129 @@ election_check(election_t *e)
 
 #define LOSS_DAMPEN 2           /* in seconds */
 
+struct vote {
+    const char *op;
+    const char *from;
+    const char *version;
+    const char *election_owner;
+    int election_id;
+    struct timeval age;
+};
+
+/*!
+ * \brief Unpack an election message
+ *
+ * \param[in] e        Election object
+ * \param[in] message  Election message XML
+ * \param[out] vote    Parsed fields from message
+ *
+ * \return TRUE if election message and election are valid, FALSE otherwise
+ * \note The parsed struct's pointer members are valid only for the lifetime of
+ *       the message argument.
+ */
+static bool
+parse_election_message(election_t *e, xmlNode *message, struct vote *vote)
+{
+    CRM_CHECK(message && vote, return FALSE);
+
+    vote->election_id = -1;
+    vote->age.tv_sec = -1;
+    vote->age.tv_usec = -1;
+
+    vote->op = crm_element_value(message, F_CRM_TASK);
+    vote->from = crm_element_value(message, F_CRM_HOST_FROM);
+    vote->version = crm_element_value(message, F_CRM_VERSION);
+    vote->election_owner = crm_element_value(message, F_CRM_ELECTION_OWNER);
+
+    crm_element_value_int(message, F_CRM_ELECTION_ID, &(vote->election_id));
+
+    if ((vote->op == NULL) || (vote->from == NULL) || (vote->version == NULL)
+        || (vote->election_owner == NULL) || (vote->election_id < 0)) {
+
+        crm_warn("Invalid %s message from %s in %s ",
+                 (vote->op? vote->op : "election"),
+                 (vote->from? vote->from : "unspecified node"),
+                 (e? e->name : "election"));
+        return FALSE;
+    }
+
+    // Op-specific validation
+
+    if (crm_str_eq(vote->op, CRM_OP_VOTE, TRUE)) {
+        // Only vote ops have uptime
+        int age_s = -1;
+        int age_us = -1;
+
+        // @TODO add functions to parse time_t / suseconds_t directly from XML
```

Comment:
`suseconds_t`?

Reply:
Signed microseconds. These are the standard types of the members of struct timeval (see gettimeofday(2)), which we're using to store the uptime.

Reply:
Aha, is not very apparent.

Reply:
Addressed in #1606